From 3a59831ac84fbcc1d02d4a3b3ecd2077bcd81ee4 Mon Sep 17 00:00:00 2001
From: David Rokhvarg <David.Rokhvarg@amd.com>
Date: Wed, 6 Jan 2016 15:49:08 -0500
Subject: [PATCH 0669/1110] drm/amd/dal: Balance dc surface reference count.

This resolves memory leak because of double increment of
surface reference count in dc_commit_surfaces_to_target().

Signed-off-by: David Rokhvarg <David.Rokhvarg@amd.com>
Acked-by: Jordan Lazare <Jordan.Lazare@amd.com>
---
 drivers/gpu/drm/amd/dal/dc/core/dc_resource.c | 5 +++--
 drivers/gpu/drm/amd/dal/dc/core/dc_target.c   | 2 +-
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/dc/core/dc_resource.c b/drivers/gpu/drm/amd/dal/dc/core/dc_resource.c
index ab081c1..1ad317a 100644
--- a/drivers/gpu/drm/amd/dal/dc/core/dc_resource.c
+++ b/drivers/gpu/drm/amd/dal/dc/core/dc_resource.c
@@ -388,8 +388,9 @@ bool logical_attach_surfaces_to_target(
 	uint8_t i;
 	struct core_target *target = DC_TARGET_TO_CORE(dc_target);
 
-	if (target->status.surface_count >= MAX_SURFACE_NUM) {
-		dal_error("Surface: this target has too many surfaces!\n");
+	if (surface_count > MAX_SURFACE_NUM) {
+		dal_error("Surface: can not attach %d surfaces! Maximum is: %d\n",
+			surface_count, MAX_SURFACE_NUM);
 		return false;
 	}
 
diff --git a/drivers/gpu/drm/amd/dal/dc/core/dc_target.c b/drivers/gpu/drm/amd/dal/dc/core/dc_target.c
index 433f712..31374ab 100644
--- a/drivers/gpu/drm/amd/dal/dc/core/dc_target.c
+++ b/drivers/gpu/drm/amd/dal/dc/core/dc_target.c
@@ -259,7 +259,6 @@ bool dc_commit_surfaces_to_target(
 					LOG_MINOR_COMPONENT_DC,
 					"0x%x:",
 					surface);
-		dc_surface_retain(surface);
 
 		program_gamma(dc->ctx, surface,
 			DC_STREAM_TO_CORE(target->public.streams[0])->ipp,
@@ -271,6 +270,7 @@ bool dc_commit_surfaces_to_target(
 
 		dc->hwss.update_plane_address(core_surface, target);
 	}
+
 	if (current_enabled_surface_count == 0 && new_enabled_surface_count > 0)
 		dc_target_enable_memory_requests(dc_target);
 
-- 
2.7.4

